File: Search in a Big Sorted Array.cpp
/**
 * Definition of ArrayReader:
 * 
 * class ArrayReader {
 * public:
 *     int get(int index) {
 *          // return the number on given index, 
 *          // return -1 if index does not exist.
 *     }
 * };
 */
class Solution {
public:
    /**
     * @param reader: An instance of ArrayReader.
     * @param target: An integer
     * @return: An integer which is the first index of target.
     */
    int searchBigSortedArray(ArrayReader *reader, int target) {
        // write your code here
        int k = 1;
        int va = reader->get(k);
        while (va != -1 && va < target)
        {
            k *= 2;
            va = reader->get(k);
        }
        int left = 0;
        int right = k;
        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            int val = reader->get(mid);
            if (val == -1 || val > target)
            {
                right = mid - 1;
            }
            else if (val < target)
            {
                left = mid + 1;
            }
            else
            {
                while (reader->get(mid) == target)
                {
                    --mid;
                }
                return mid + 1;
            }
        }
        return -1;
        
    }
};

File: Search in a Big Sorted Array.py
"""
Definition of ArrayReader:
class ArrayReader:
    def get(self, index):
        # this would return the number on the given index
        # if there is no number on that index, return -1
"""
class Solution:
    # @param {ArrayReader} reader: An instance of ArrayReader 
    # @param {int} target an integer
    # @return {int} an integer
    def searchBigSortedArray(self, reader, target):
        # write your code here
        
        if not reader:
            return -1
        
        k = 1
        while reader.get(k) != -1 and reader.get(k) < target:
            k *= 2
        left, right = 0, k
        while left + 1 < right:
            mid = left + (right - left) / 2
            if reader.get(mid) == -1 or reader.get(mid) >= target:
                right = mid
            elif reader.get(mid) < target:
                left = mid

        if reader.get(left) == target:
            return left
        if reader.get(right) != -1 and reader.get(right) == target:
            return right
        return -1